### DISCUSSION TITLE
Get Voronoi diagram

### DISCUSSION THREAD

**Discussion by Joseph-HLLin** [BrunoLevy/geogram#100] [Q&A] [answered]
Hi developers,
I'd like to get the Voronoi diagram of a surface mesh from RVD function.  However, when I used "compute_RVD", it gave me the intersection of mesh and Voronoi diagram.  Does anyone know how to get the Voronoi diagram only?
Thank you so much : )

**Comment by BrunoLevy** [comment, marked as answer]
Hi,
"compute_RVD" computes the intersection of two meshes:
- the Voronoi diagram of the points
- the surface mesh or volume mesh this Voronoi diagram is restricted to
If I understood well, what you want is "throwing away" the vertices yielded by the intersections between the Voronoi cells and the mesh vertices, and only keeping the Vornoi vertices. Depending on whether you are in 2D, in 3D restricted to a surface or in 3D restricted to a volume, it will be different, but the general idea is as follows:
For each seed:
- traverse all the polygons / polyhedra that correspond to the restricted Voronoi cell associated with the seed
- traverse the border of the restricted Voronoi cell, and discard the vertices that you do not want

Here is the command to generate a (volumetric) Voronoi diagram restricted to a tet mesh:
```
  $ vorpalite profile=poly pts=300 holes.obj
```
![image](https://github.com/BrunoLevy/geogram/assets/19478253/72ec063e-c59a-43c0-8b7a-873d56b3a388)

As can be seen, you get on the boundary all the intersections between the Voronoi cells and the triangles of the surface, which is normal, because they can be non-planar (so in general you want them).

You can merge some of these polygons based on the angles between their normals (merge them if the angle is smaller than the specified threshold), as follows:
```
  $ vorpalite profile=poly pts=300 holes.obj poly:normal_angle_threshold=5 poly:simplify=tets_voronoi_boundary
```
Then you obtain this result:
![image](https://github.com/BrunoLevy/geogram/assets/19478253/6e05de03-cf6f-40a6-8298-9c35faa757c0)

To see how to do that programatically, refer to the [sources of voraplite](https://github.com/BrunoLevy/geogram/blob/1d14ae91b59bf33ca308462bf4fa2c66d9d85215/src/bin/vorpalite/main.cpp#L358) and to the [documentation of the CentroidalVoronoiTessellation class](https://brunolevy.github.io/geogram/classGEO_1_1CentroidalVoronoiTesselation.html)
